README.txt

= ruote-rest

A RESTful instance of OpenWFEru (ruote) powered by Rack (http://rack.rubyforge.org/)


== getting it

Get it from GitHub or download a prepackaged release at http://rubyforge.org/frs/?group_id=2609 (and jump to '== preparing it')

To get Ruote and Ruote-Rest :

  git clone git://github.com/jmettraux/ruote-rest.git

Then

  cd ruote-rest 
  rake ruote:install

to install the Ruote (OpenWFEru) workflow engine and its dependencies (in a vendor/ subdirectory). Alternatively, you could do 

  rake ruote:gem_install

to get the dependencies as gems.

If you want to grab the gem and freeze them under vendorf/ do :

  rake ruote:install_freeze


== preparing it

To prepare the development database

  rake mysql:setup

Expects a mysql db with a 'root' admin account with sufficient access rights. 
It will create a database named "ruoterest_development".

(
To prepare the test database

  rake mysql:setup stage=test

To prepare the dev database with the admin 'toto'

  rake mysql:setup dbadmin=toto
)

The last step of the database preparation is loading the authentication settings :

  rake mysql:populate

These settings may be modified by editing tasks/fixtures/host.yml and/or tasks/fixtures/users.yml

To generate a password for a user :

  rake password:generate smd5 my_very_secret_password

or

  rake password:generate ssha my_very_secret_password

The resulting string (something like "{SMD5}HKBKsOPQ1PleLG3KOlmHTWtoNW9HVGxC") can be inserted in the fixture or in the 'password' column database for the given user.


== starting it

  ruby lib/start.rb


Then head to

  http://localhost:4567/

(
to start it on port 3333 :

  ruby lib/start.rb -p 3333
)

it will lead you to the "service document" with links to all the resources/
collection that make ruote-rest.


== interface

Just navigate the interface with your browser, everything is there.


== configuration

conf/db.rb

  database configuration

conf/engine.rb

  engine configuration

conf/participants.rb
conf/participants_development.yaml

  participants configuration, the yaml file holds the list of 
  'active participants' (the worklist in fact).

conf/auth.rb

  authentication filters, contains an HTTP basic authentication example
  and a "whitelist" authentication example.


== license

MIT


== feedback

user mailing list :        http://groups.google.com/group/openwferu-users
developers mailing list :  http://groups.google.com/group/openwferu-dev

issue tracker :            http://github.com/jmettraux/ruote-rest/issues/

irc :                      irc.freenode.net #ruote



CHANGELOG.txt

= ruote-rest CHANGELOG.txt


== ruote-rest - 0.9.21    not yet released

- issue    : with on_cancel and get /processes. Thanks Gonzalo
- issue    : render_fei_json missing (Thanks Gonzalo)
- issue #5 : expressions/workitems in subprocesses not visible. (Thanks Patrick)
- issue #4 : unlocked full range of db params for activerecord (Thanks Nando)
- todo  #25167 : rake password:generate
- todo  #25799 : post /expressions/:wfid/:expid linked to engine.reapply(exp)
- todo  #25858 : reply page for things like "process x got cancelled"
- todo  #25710 : made workitem [payload] editable at error replay time
- todo  #25167 : integrated Gonzalo and Nando's newauth patch
- bug   #25016 : gem 'activerecord' getting in the way of frozen.rb. Removed.
- todo  #23769 : moved to fs_engine
- todo  #23869 : adapted to new ProcessStatus#workitem_expids
- todo  #23851 : implemented missing #parse_workitem_json
- todo  #23850 : making sure to reply 200 when deleting a res over json or xml
- todo  #23600 : adapted to file renaming in ruote itself
- bug   #23580 : engine.reload performed before participants are ready. Fixed.
- todo  #23226 : fixed path issue with daemons (Kenneth Kalmer)
- todo  #23041 : daemonize patch (Kenneth Kalmer)
- todo  #22966 : enhanced authentication (Kenneth Kalmer)
- bug   #22920 : no participant at /participants/.* fixed.
- todo  #22919 : now emitting 400 when failing to parse incoming representation
- todo  #22918 : workitems/:wfid instead of workitems?wfid=:wfid
- todo  #22890 : error_to_json integrated
- todo  #22898 : expression_to_json integrated
- todo  #22864 : expression_to_xml moved to ruote itself
- todo  #22868 : error_to_xml moved to ruote itself
- todo  #22866 : /workitems/:wfid/:expid is now the norm
- todo  #22834 : adapted to new ruote:lib/openwfe/representations.rb
- todo  #22699 : expressions/X/Y/raw -> expressions/X/Y/tree
- todo  #22005 : moved process_to_xml to ruote (openwfe/representations.rb)
- todo  #20613 : cleaned up xml generation (builder(options) do |xml|)
- todo  #21963 : 'tree' replaces 'representation'
- todo  #21952 : wrote test/all_reps.rb (all representations on the deck)


== ruote-rest - 0.9.19    released 2008/09/09

- todo  #21841 : wired in raw expression [fluo] edition
- todo  #19967 : etag and last-modified for everybody
- todo  #21707 : integrated atom feed for /history
- todo  #21722 : replaced TreeChecker by the rufus-treechecker gem
- todo  #21646 : now taking advantage of the process status cache in the engine
- todo  #21644 : included process_stack result into process_status
- todo  #21638 : added the /service 'document'
- todo  #21509 : added /history view
- todo  #21598 : better workitem ids (backward compatible though)
- todo  #21575 : adapted to sixjo params['_FORMAT'] (.xml, .json, ...)
- todo  #21413 : showing workitem search form
- patch #21497 : participant initialization parameter fix by Nick Petrella
- todo  #21419 : removed direct dependency on json gem, relying on whatever is 
                 present (activesupport or json)
- bug   #21096 : reordered participant load order in conf/participants.rb
- todo  #20647 : showing scheduled jobs for /processes



lib/inout.rb
#--
# Copyright (c) 2008-2009, John Mettraux, jmettraux@gmail.com
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
# Made in Japan.
#++


class Rufus::Sixjo::Context

  # Parses the representation sent in the request body
  #
  # If tolerant is set to true, will simply return nil when it fails to parse
  # the incoming representation.
  #
  def rparse (type, tolerant=false)

    request.body.rewind
    representation = request.body.read

    format = determine_in_format

    begin
      #p [ "parse_#{type}_#{format}", representation ]
      send("parse_#{type}_#{format}", representation)
    rescue Exception => e
      return nil if tolerant
      throw :done, [ 400, "failed to parse incoming representation" ]
    end
  end

  # The entry point for rendering any ruote-rest object
  #
  # (pronounce with a "rolling r")
  #
  def rrender (type, object, options={})

    format, ctype = determine_out_format(options)

    ctype = 'text/plain' if params[:plain]
      # useful for debugging

    response.status = options.delete(:status) || 200
    response.content_type = ctype

    options.each { |k, v| response.header[k] = v }

    set_etag(object.etag) \
      if object.respond_to?(:etag) and object.etag
    set_last_modified(object.timestamp) \
      if object.respond_to?(:timestamp) and object.timestamp
        #
        # where the conditional GET happens...


    unless format == 'js'

      body = send("render_#{type}_#{format}", object)

      body = body.gsub(', ', ",\n ") \
        if format == 'json' and ctype == 'text/plain'
          # a bit more readable for 'text/plain' output

      return body
    end

    varname = params[:var] || 'ruote_js'
    method = "render_#{type}_json"

    "var #{varname} = #{send(method, object)}"
  end

  # Simply reads the "Content-Type" header
  #
  def determine_in_format

    ct = request.env['CONTENT_TYPE'] || ''

    return 'form' if ct.index('form-')
    return 'json' if ct.index('application/json')
    return 'yaml' if ct.index('application/yaml')

    'xml'
  end

  # some common formats
  #
  FORMATS = {

    :xml => [ 'xml', 'application/xml' ],
    :html => [ 'html', 'text/html' ],
    :json => [ 'json', 'application/json' ],
    :js => [ 'js', 'text/javascript' ],
    :yaml => [ 'yaml', 'application/yaml' ],
    :atom => [ 'atom', 'application/atom+xml' ]

  } unless defined?(FORMATS)

  FTYPES = FORMATS.keys.collect { |k| k.to_s } \
    unless defined?(FTYPES)

  # Determines the format the client is expecting
  #
  def determine_out_format (options={})

    f = options[:format] || params[:format] || request.env['_FORMAT']

    return FORMATS[:xml] if f == 'xml'
    return FORMATS[:json] if f == 'json'
    return FORMATS[:js] if f == 'js'
    return FORMATS[:yaml] if f == 'yaml'
    return FORMATS[:atom] if f == 'atom'

    accept = request.env['HTTP_ACCEPT'] || ''

    return FORMATS[:html] if accept.index('text/html')
    return FORMATS[:yaml] if accept.index('yaml')
    return FORMATS[:json] if accept.index('json')
    return FORMATS[:js] if accept.index('js')
    return FORMATS[:atom] if accept.index('atom')

    FORMATS[:xml]
  end

  # Returns true if the string is something like "xxx.json" or "yyy.xml"
  #
  def has_filetype? (s)

    ss = s.split('.')
    return false if ss.length != 2

    FTYPES.include?(ss.last)
  end

end

